### DISCUSSION TITLE
Crazyswarm installation issues

### DISCUSSION THREAD

**Discussion by DavidGMcIntosh** [USC-ACTLab/crazyswarm#399] [Q&A]
I'm attempting to follow the setup instructions at https://crazyswarm.readthedocs.io/en/latest/installation.html#installation, but I am reaching problems at step 3. ROS is installed and fully functional, but when attempting to install the dependencies, I always get an error message stating that it is 'Unable to locate package gcc-arm-embedded'. I have tried to install this dependency elsewhere, but no matter what I do to download it I am met with the same error. I have tried ignoring this error, but although I am still able to run the ./build.sh command, the tests to verify it  fail ('$CSW_PYTHON -m pytest' is met with "No module named pytest"). Does anyone know a way to use crazyswarm without this dependency, a foolproof way to install it, or something else that may be causing this issue? Thanks! 
[Pictures of errors attached, using Xubuntu 20.04, ROS Noetic]
![Screen Shot 2021-06-22 at 5 31 41 PM](https://user-images.githubusercontent.com/68243243/123003337-0f8be380-d381-11eb-87b2-2685401b2457.png)
![Screen Shot 2021-06-22 at 5 27 36 PM](https://user-images.githubusercontent.com/68243243/123003329-0dc22000-d381-11eb-9b05-e37d9ea19a76.png)
![Screen Shot 2021-06-22 at 5 27 52 PM](https://user-images.githubusercontent.com/68243243/123003341-1155a700-d381-11eb-8c71-1ad3c7032a08.png)

**Comment by jpreiss** [comment]
You don't need to rebuild the firmware unless you have modified it. We will remove that from `build.sh` eventually. See https://github.com/USC-ACTLab/crazyswarm/discussions/391

**Reply by DavidGMcIntosh** [reply]
Does this mean gcc-arm-embedded is not necessary? I'm still unable to run the pytest script, which I assumed meant the installation was in some way unsuccessful. Is there a different fix for this?

**Reply by jpreiss** [reply]
Ah, you may need to install the Ubuntu package `python3-pytest`. This is a docs bug.

**Reply by DavidGMcIntosh** [reply]
Alright, I was able to do that, and it runs. When I do, though, it gives 28 failed test errors that all read ImportError: cannot import name 'cffirmware' from 'pycrazyswarm.cfsim' (/home/bitcraze/crazyswarm/ros_ws/src/crazyswarm/scripts/pycrazyswarm/cfsim/__init__.py). 
I tried looking into this, but the __init__.py file in question appears to be empty? I'm pretty new at this, so my apologies if these questions have obvious answers, but it wasn't clear to me what should be done to proceed.

**Reply by jpreiss** [reply]
You still need to run this step:

    # build simulator firmware backend
    cd ros_ws/src/crazyswarm/scripts/pycrazyswarm/cfsim
    make

**Reply by DavidGMcIntosh** [reply]
Okay - the make command worked (though I needed to install swig first, if this is worth documenting) but I have a different error now. There are still 28 failed tests, only now the error looks something like this [see image]. It isn't as clear where the problem is, but I tried connecting a controller with a joystick and running it again, which did not resolve the issue. Any thoughts?
![Screen Shot 2021-06-23 at 4 35 25 PM](https://user-images.githubusercontent.com/68243243/123166612-9b664400-d443-11eb-903a-10ab54b628a4.png)

**Reply by jpreiss** [reply]
It looks like we made an invalid assumption in `ros_ws/src/crazyswarm/scripts/pycrazyswarm/genericJoystick.py`. Please try replacing the body of the `try` block in `Joystick.__init__()` https://github.com/USC-ACTLab/crazyswarm/blob/7c2aaeb868415b19fe5c22ac2e2c405d65f314da/ros_ws/src/crazyswarm/scripts/pycrazyswarm/genericJoystick.py#L33-L41 with the following code:

```python
        try:
            from . import linuxjsdev
            self.js = linuxjsdev.Joystick()
            devices = self.js.devices()
            if len(devices) == 0:
                print("Warning: No joystick found!")
            else:
                self.hasJoystick = True
                joyID = devices[0]["id"]
                self.js.open(joyID)
```
and report back if it fixes the problem.

**Reply by jpreiss** [reply]
`swig` is already in the list of dependencies in the "Physical robots and simulation" tab of https://crazyswarm.readthedocs.io/en/latest/installation.html.

**Reply by DavidGMcIntosh** [reply]
28 passed! Thank you, that worked! 
[Also, you're totally right on swig, I made a mistake there]

**Reply by jpreiss** [reply]
thanks for reporting back!

**Reply by DavidGMcIntosh** [reply]
Also, out of curiosity, was the section in joystick.__init__() a bug, or a workaround for something else I set up wrong? Thanks again

**Reply by jpreiss** [reply]
It was a bug. We assumed that if you have joystick devices, one of them has the id `0`. Apparently, on your system this assumption is wrong.

**Reply by jpreiss** [reply]
The patch I suggested earlier is enough to pass tests, but it does not fix the full problem when you run with real hardware. The complete fix is in https://github.com/USC-ACTLab/crazyswarm/pull/404. Please wait for it to pass CI & review and be merged before you try to run Crazyswarm on real hardware. (Or switch from `USC-ACTLab:master` to `jpreiss:joystick-id` if you are in a hurry)

**Reply by DavidGMcIntosh** [reply]
Installed the new patch, everything seems functional so far. Finished setup and began configuration, and I'm currently stuck on step 2.2. Now, when i try to run the given code it tells me that there is no such code as 'prebuilt/cradio.bin', but I can't figure out where this would be installed from. Any thoughts?
![Screen Shot 2021-06-24 at 3 36 06 PM](https://user-images.githubusercontent.com/68243243/123323144-1fcbcc00-d503-11eb-84a7-cd866a714990.png)

**Reply by jpreiss** [reply]
Just now we removed the crazyflie2-nrf-firmware and crazyradio-firmware submodules from the Crazyswarm repo since we no longer use custom versions. But it looks like to upgrade the radio firmware, you will need to follow the instructions at https://www.bitcraze.io/documentation/repository/crazyradio-firmware/master/building/programming/ which still require cloning the firmware repository yourself.

**Reply by jpreiss** [reply]
@whoenig this seems kind of clunky. I was hoping the GUI client had a tool for updating radio firmware, but apparently not.

**Reply by DavidGMcIntosh** [reply]
I'm a little confused - does this mean it's alright to work with a version 0.53 Crazyradio, or should I still try to upgrade to version 99.55?

**Reply by whoenig** [reply]
@jpreiss Upgrading the radio is not as safe as the other part since it is possible to destroy the bootloader (recovery in this case requires some soldering). That's why there is no prominent GUI tool.
@DavidGMcIntosh You will need to work with an upgraded radio firmware.
